src/UCTSearch.cpp

```diff
@@ -216,16 +264,26 @@ SearchResult UCTSearch::play_simulation(GameState & currstate,
         auto next = node->uct_select_child(color, node == m_root.get());
         auto move = next->get_move();
 
+        // myprintf("%s %d ", currstate.move_to_text(move), thread_num); //// show thread number as well..
+
         currstate.play_move(move);
         if (move != FastBoard::PASS && currstate.superko()) {
             next->invalidate();
         } else {
-            result = play_simulation(currstate, next);
+            result = play_simulation(currstate, next, thread_num);
         }
     }
 
     if (result.valid()) {
-        node->update(result.eval());
+        auto eval = result.eval();
+        auto num = 1;
+        while (eval >= 2.0f) {
```

Comment:
I understand this is a result of actually doing two evaluations to get to the proper side to move. I'd try to refactor the double expansion into the node evaluation itself, and then just assign the average of both evaluations to the backup, this keeping all the existing tree search code that deals with this unaware of what happened underneath.